opentelemetry-sdk/src/opentelemetry/sdk/metrics/_internal/aggregation.py

```diff
@@ -171,16 +174,18 @@ def collect(
             # Output DELTA for a synchronous instrument
             # Output CUMULATIVE for an asynchronous instrument
             self._previous_point = current_point
+            self._previous_value = current_value
             return current_point
 
         if aggregation_temporality is AggregationTemporality.DELTA:
             # Output temporality DELTA for an asynchronous instrument
-            value = current_point.value - self._previous_point.value
+            current_value = current_point.value - self._previous_value
+            self._previous_value = current_point.value
```

Comment:
I don't really understand what's different here, it looks you added a new member called `self._previous_value`, however that was already saved in `self._previous_point.value`.

How is `self._previous_value` supposed to differ from `self._previous_point.value`?

Reply:
It is not obvious during the first collection cycle, but it matters later.

Short explanation:

`self._previous_value` is always the value of the previous measurement. `self._previous_point.value` is the value of the _difference_ between the current measurement value and the previous measurement value.

Previously we calculated `self._previous_point.value` as the difference between the current measurement value and `self._previous_point.value`, but `self._previous_point.value` is already a _difference_, leading to wrong results.

Long explanation:

We start running the test case added in this PR where we call `aggregate` with a value of `8`, then we call `collect`, then we call `aggregate` with a value of `16`, then we call `collect`, then we call `aggregate` with a value of `24` and so on.

After the `_SumAggregation` object is created:
```
ipdb> self._previous_point is None
True
ipdb> self._current_value is None
True
ipdb> self._previous_value is None
True
```

After calling `aggregate` for the first time with a value of `8`:
```
ipdb> self._previous_point is None
True
ipdb> self._current_value
8
ipdb> self._previous_value is None
True
```

After calling `collect` for the first time:
```
ipdb> self._previous_point
NumberDataPoint(attributes={}, start_time_unix_nano=1692699690552255929, time_unix_nano=1692700159102147033, value=8)
ipdb> self._current_value is None
True
ipdb> self._previous_value
8
```

At this moment both `self._previous_point.value` and `self._previous_value` are `8`.

After calling `aggregate` for the second time with a value of `16`:
```
ipdb> self._previous_point
NumberDataPoint(attributes={}, start_time_unix_nano=1692699690552255929, time_unix_nano=1692700159102147033, value=8)
ipdb> self._current_value
16
ipdb> self._previous_value
8
```

After calling `collect` for the second time:
```
ipdb> self._previous_point
NumberDataPoint(attributes={}, start_time_unix_nano=1692700159102147033, time_unix_nano=1692700700773894084, value=8)
ipdb> self._current_value is None
True
ipdb> self._previous_value
16	
```

Now `self._previous_point.value` is `8` but `self._previous_value` is `16`.

After calling `aggregate` for the third time with a value of `24`:
```
ipdb> self._previous_point
NumberDataPoint(attributes={}, start_time_unix_nano=1692700159102147033, time_unix_nano=1692700700773894084, value=8)
ipdb> self._current_value
24
ipdb> self._previous_value
16
```

After calling `collect` for the third time:
```
ipdb> self._previous_point
NumberDataPoint(attributes={}, start_time_unix_nano=1692700700773894084, time_unix_nano=1692702465661881982, value=8)
ipdb> self._current_value is None
True
ipdb> self._previous_value
24
```

As you can see, the `NumberDataPoint` is always `8`, which is right because it is the difference between the current measurement value and the previous measurement value. To be able to calculate this difference we need the actual previous measurement value, not the difference between the previous measurement value and the previous previous measurement value.

Reply:
> `self._previous_value` is always the value of the previous measurement. `self._previous_point.value` is the value of the _difference_ between the current measurement value and the previous measurement value.

I see, that sounds like the actual bug. From the original design, the `self.previous_point` should ALWAYS be the cumulative value up until the last collection. Then it is straightforward to calculate either CUMULATIVE or DELTA. Copying from a previous discussion:

> ```
> time --->
> 
> previous_point_cumulative:
> |-----------------|
> 
> current_point_cumulative:
> |-------------------------|
> 
> current_point_delta:
>                   |-------|
> ```
> 
> Given you saved `previous_point_cumulative` from the previous collection:
> - If you have `current_point_delta` (synchronous instrument), you can generate `current_point_cumulative = previous_point_cumulative + current_point_delta`
> - If you have `current_point_cumulative` (async instrument), you can generate `current_point_delta = current_point_cumulative - previous_point_cumulative`
> 
> Check out the [Temporality Conversion section](https://docs.google.com/document/d/1FOmpGCiZAWTd6a3vcEHBnm1LKp7HAKuD4BBN3YCkymc/edit#heading=h.ulghopzfw7ou) of the doc too

Reply:
Fixed, it uses that approach now.